internal/infrastructure/ceph/rbd_test.go

```diff
@@ -517,11 +517,317 @@ func TestApplyDiffToRawImage_error_OverlappedDataRecords(t *testing.T) {
+		}),
+	)
+	require.NoError(t, err)
+
+	blockDevicePath := getBlockDevicePathForTest(t)
+	zerooutWholeBlockDevice(t, blockDevicePath)
+
+	// Act
+	err = applyDiffToBlockDevice(blockDevicePath, reader, "", "toSnap")
+
+	// Assert
+	assert.NoError(t, err)
+
+	file := openFile(t, blockDevicePath)
+	head := make([]byte, 30)
+	_, err = io.ReadFull(file, head)
+	require.NoError(t, err)
+	assert.Equal(
+		t,
+		// UPDATED DATA ("0123456789") + ZERO DATA (20 bytes)
+		[]byte("0123456789\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00"+
+			"\x00\x00\x00\x00\x00\x00\x00\x00\x00"),
+		head,
+	)
+
+	// Ensure the rest of the block device is zeroed out
+	compareReaders(t, file, io.LimitReader(&zeroReader{}, int64(getBlockDeviceSize(t, file)-30)))
+}
+
+func TestApplyDiffToBlockDevice_success_VariousZeroDataRecords(t *testing.T) {
+	// Description:
+	// For ZERO DATA, verify that the region offset ~ offset + length is overwritten with 0 on the block device
+	//
+	// Arrange:
+	// There exists an incremental data file containing a single ZERO DATA record,
+	// and that record satisfies one of the following conditions:
+	// - Both offset and offset + length are aligned, and the range offset ~ offset + length
+	//   includes an entire sector.
+	// - Offset is aligned, offset + length is not aligned, and the range offset ~ offset + length
+	//   includes an entire sector.
+	// - Offset is aligned, offset + length is not aligned, and the range offset ~ offset + length
+	//   does not include an entire sector.
+	// - Offset is not aligned, offset + length is aligned, and the range offset ~ offset + length
+	//   includes an entire sector.
+	// - Offset is not aligned, offset + length is aligned, and the range offset ~ offset + length
+	//   does not include an entire sector.
+	// - Neither offset nor offset + length are aligned, and the range offset ~ offset + length
+	//   includes an entire sector.
+	// - Neither offset nor offset + length are aligned, and the range offset ~ offset + length
+	//   does not include an entire sector.
+	//
+	// Act:
+	// Call the apply process using the incremental data file
+	//
+	// Assert:
+	// - Completes successfully
+	// - The region offset ~ offset + length of the block device is overwritten with 0
+
+	blockDevicePath := getBlockDevicePathForTest(t)
+	file := openFile(t, blockDevicePath)
+	blockDeviceSectorSize := getBlockDeviceSectorSize(t, file)
+
+	testCases := []struct {
+		name                 string
+		zeroDataRecordOffset uint64
+		zeroDataRecordLength uint64
+		expected             []byte
+	}{
+		{
+			name:                 "aligned offset and (offset + length)",
+			zeroDataRecordOffset: uint64(blockDeviceSectorSize),
+			zeroDataRecordLength: uint64(blockDeviceSectorSize),
+			expected: bytes.Join(
+				[][]byte{
+					bytes.Repeat([]byte{0xff}, blockDeviceSectorSize),
+					bytes.Repeat([]byte{0x00}, blockDeviceSectorSize),
+					bytes.Repeat([]byte{0xff}, blockDeviceSectorSize),
+				},
```

Comment:
We don't have to construct expected binary actually. What we need is information about "where's the (successive) zero range?", so something like `expectedOffset: blockDeviceSectorSize, expectedLength: blockDeviceSectorSize` would be enough.

Reply:
Maybe this is not a good idea.